### PR TITLE
Added callback TerminateOnNaN() to terminate training if NaN loss enc…

### PR DESCRIPTION
**Pull request by cengh**
…ountered in batch

In cases where jobs are submitted to a queue, it would be helpful to have training terminated automatically when NaN loss is encountered, as stated in issue #4848. This PR uses a callback, which is applied to every Keras model, to stop training at the end of a batch if the loss is NaN.

### CODE REVIEW

**Comment by patyork** [comment]
At the very least, this should be a check for all invalid values, including `inf` values.

In addition, forcing a stop of training without a (re)actionable exception is just as bad as not stopping at all. Unless you're watching the training all the time, simply forcing a stop of training without providing a way to acknowledge the issue and rectify it before automatically continuing does nothing but save some electricity. As an example, my nan-able models use a callback to check (same as this) but provides for the ability to drop a minibatch and/or reload weights from a previous epoch and continuing from there; all done with no intervention on my part unless something very bad happens and it goes into a loop of NaN->reload weights->NaN.

**Comment by cengh** [comment]
For my use cases, when an invalid loss is encountered, I do actually just want to stop training the current model in order to free up the computational resources to train a different model. For instance, if I am performing an automated hyperparameter search, I just want to reject poor sets of hyperparameters quickly.

I have added a check for inf values, and a patience parameter to permit multiple consecutive mini-batches with an invalid loss. However, it seems to me that once training of the current model has been terminated, whether to a) restart training the same model with 'modifications', such as reverting to weights from a previous epoch or b) start training a new model entirely, is a separate question. In any case, it seems like restarting training of the same model would be better implemented by another script that manages automatic job submission.

**Comment by hubayirp** [comment]
+1

**Comment by fchollet** [comment]
We will not be merging new APIs for the time being.

**Comment by cengh** [comment]
@fchollet 

I've updated the implementation so that the `TerminateOnNaN()` callback is no longer applied to every model. 

The modifications in keras/engine/training.py just wrap the nested for loops (iterating over epochs and batches) in `_fit_loop()` and `fit_generator()` in try/except statements. The try/except statements allow a check on callback_model.stop_training to be performed at the end of a batch of training, instead of having to wait until the end of an epoch. This is helpful when working with large datasets where a single epoch of training takes a long time. 

In any case, it could be useful to have a `TerminateOnNaN()` callback that is capable of terminating training after an epoch of training where the loss is NaN (checking every epoch instead of every batch). This would not require any changes to keras/engine/training.py. What do you think?

**Comment by kielnino** [comment]
Why not use a custom callback?
```
import keras
import math


class NANCallback(keras.callbacks.Callback):
    def on_batch_end(self, batch, logs={}):
        loss = logs.get('loss')
        if math.isnan(loss):
            self.model.stop_training = True
```
